File: examples/apps/cali-perfproblem-branch-mispred.cpp
// Copyright (c) 2019, Lawrence Livermore National Security, LLC.
// See top-level LICENSE file for details.

#include <iostream>
#include <cstdlib>
#include <algorithm>

#include <caliper/cali.h>
#include <caliper/cali_datatracker.h>

int* init(size_t arraySize, bool sort)
{
    CALI_CXX_MARK_FUNCTION;

    int *data = 
        static_cast<int*>(cali_datatracker_allocate_dimensional("data", sizeof(int), &arraySize, 1));

    std::srand(1337);
    for (size_t c = 0; c < arraySize; ++c)
        data[c] = std::rand() % 256;

    if (sort)
        std::sort(data, data + arraySize); 

    return data;
}

void work(int *data, size_t arraySize)
{
    CALI_CXX_MARK_FUNCTION;

    long sum = 0;

    for (size_t i = 0; i < 100000; ++i)
    {
        // Primary loop
        for (size_t c = 0; c < arraySize; ++c)
        {
            if (data[c] >= 128)
                sum += data[c];
        }
    }
    std::cout << "sum = " << sum << std::endl;
}

void cleanup(int *data)
{
    CALI_CXX_MARK_FUNCTION;

    cali_datatracker_free(data);
}

void benchmark(size_t arraySize, bool sort)
{
    CALI_CXX_MARK_FUNCTION;

    cali::Annotation sorted("sorted");
    sorted.set(sort);

    std::cout << "Initializing benchmark data with sort = " << sort << std::endl;

    int *data = init(arraySize, sort);

    std::cout << "Calculating sum of values >= 128" << std::endl;

    work(data, arraySize);

    std::cout << "Cleaning up" << std::endl;

    cleanup(data);

    std::cout << "Done!" << std::endl;
}

int main(int argc, char *argv[])
{
    CALI_CXX_MARK_FUNCTION;

    // Generate data
    size_t arraySize = 32768;
    benchmark(arraySize, true);
    benchmark(arraySize, false);
}

File: examples/apps/cali-perfproblem-branch-mispred.py
# Copyright (c) 2024, Lawrence Livermore National Security, LLC.
# See top-level LICENSE file for details.

from pycaliper.high_level import annotate_function
from pycaliper.annotation import Annotation

import numpy as np

@annotate_function()
def init(arraySize: int, sort: bool) -> np.array:
    data = np.random.randint(256, size=arraySize)
    if sort:
        data = np.sort(data)
    return data


@annotate_function()
def work(data: np.array):
    data_sum = 0
    for _ in range(100):
        for val in np.nditer(data):
            if val >= 128:
                data_sum += val
    print("sum =", data_sum)


@annotate_function()
def benchmark(arraySize: int, sort: bool):
    sorted_ann = Annotation("sorted")
    sorted_ann.set(sort)
    print("Intializing benchmark data with sort =", sort)
    data = init(arraySize, sort)
    print("Calculating sum of values >= 128")
    work(data)
    print("Done!")
    sorted_ann.end()
    

@annotate_function()
def main():
    arraySize = 32768
    benchmark(arraySize, True)
    benchmark(arraySize, False)
    

if __name__ == "__main__":
    main()
    
